File: Daemon/Asn1ServiceBindings.cs
using Daemon.Utils;
using Service;
using Service.Dto;
using System.Net;

namespace Daemon
{
    class Asn1ServiceBindings
    {
        public Asn1ServiceBindings(IAsn1Service service)
        {
            this.service = service;
        }

        public void BindTo(HttpServer server)
        {
            server.InstallHandler("version", GetVersion);
            server.InstallHandler("ast", BuildAst);
        }

        private void GetVersion(HttpListenerContext context)
        {
            context.Response.SendPlainText(service.GetVersion());
        }

        private void BuildAst(HttpListenerContext context)
        {
            var data = context.Request.ReadJson<InputFiles>();
            context.Response.SendJson(service.BuildAst(data));
        }

        private IAsn1Service service;
    }
}

File: Daemon/Asn1ServiceBindings.fs
namespace Daemon

open Service
open System.Net

type Asn1ServiceBindings(service:IAsn1Service) =

    member this.GetVersion (context:HttpListenerContext) = Http.SendPlainText context.Response service.Version

    member this.BuildAst (context:HttpListenerContext) = 
        let data = Http.ReadJson<Dto.InputFiles> context.Request
        Http.SendJson context.Response (service.BuildAst data)

